lib/Frontend/ParseableInterfaceSupport.cpp

```diff
@@ -0,0 +1,332 @@
+  for (const ModuleDecl::ImportedModule &importPair : imports) {
+    if (importPair.first.empty())
+      continue;
+    diags.diagnose(importPair.first.front().second,
+                   diag::parseable_interface_scoped_import_unsupported);
+  }
+}
+
+/// Prints to \p out a comment containing a tool-versions identifier as well
+/// as any relevant command-line flags in \p Opts used to construct \p M.
+static void printToolVersionAndFlagsComment(raw_ostream &out,
+                                            ParseableInterfaceOptions const &Opts,
+                                            ModuleDecl *M) {
+  auto &Ctx = M->getASTContext();
+  out << "// " SWIFT_TOOLS_VERSION_KEY ": "
+      << Ctx.LangOpts.EffectiveLanguageVersion << "\n";
+  out << "// " SWIFT_MODULE_FLAGS_KEY ": "
+      << Opts.ParseableInterfaceFlags << "\n";
+}
+
+llvm::Regex swift::getSwiftInterfaceToolsVersionRegex() {
+  return llvm::Regex("^// " SWIFT_TOOLS_VERSION_KEY ": ([0-9\\.]+)$",
+                     llvm::Regex::Newline);
+}
+
+llvm::Regex swift::getSwiftInterfaceModuleFlagsRegex() {
+  return llvm::Regex("^// " SWIFT_MODULE_FLAGS_KEY ": (.*)$",
+                     llvm::Regex::Newline);
+}
+
+/// Prints the imported modules in \p M to \p out in the form of \c import
+/// source declarations.
+static void printImports(raw_ostream &out, ModuleDecl *M) {
+  // FIXME: This is very similar to what's in Serializer::writeInputBlock, but
+  // it's not obvious what higher-level optimization would be factored out here.
+  SmallVector<ModuleDecl::ImportedModule, 8> allImports;
+  M->getImportedModules(allImports, ModuleDecl::ImportFilter::All);
+  ModuleDecl::removeDuplicateImports(allImports);
+  diagnoseScopedImports(M->getASTContext().Diags, allImports);
+
+  // Collect the public imports as a subset so that we can mark them with
+  // '@_exported'.
+  SmallVector<ModuleDecl::ImportedModule, 8> publicImports;
+  M->getImportedModules(publicImports, ModuleDecl::ImportFilter::Public);
+  llvm::SmallSet<ModuleDecl::ImportedModule, 8,
+                 ModuleDecl::OrderImportedModules> publicImportSet;
+  publicImportSet.insert(publicImports.begin(), publicImports.end());
+
+  for (auto import : allImports) {
+    if (import.second->isStdlibModule() ||
+        import.second->isOnoneSupportModule() ||
+        import.second->isBuiltinModule()) {
+      continue;
+    }
+
+    if (publicImportSet.count(import))
+      out << "@_exported ";
+    out << "import ";
+    import.second->getReverseFullModuleName().printForward(out);
+
+    // Write the access path we should be honoring but aren't.
+    // (See diagnoseScopedImports above.)
+    if (!import.first.empty()) {
+      out << "/*";
+      for (const auto &accessPathElem : import.first)
+        out << "." << accessPathElem.first;
+      out << "*/";
+    }
+
+    out << "\n";
+  }
+}
+
+bool swift::emitParseableInterface(raw_ostream &out,
```

Comment:
I think I'd prefer to keep this in a separate file from the loader, but I don't feel too strongly about it.